iree/turbine/kernel/ops/wave_ops.py

```diff
@@ -1030,6 +1037,52 @@ def index(self, value: dict[IndexSymbol, IndexSequence]):
         CustomOp.index.fset(self, value)
 
 
+@define_op("extract")
+@dataclass
+class Extract(CustomOp):
+    """
+    Op Rationale:
+
+    Extract is an op used to represent extracting of
+    a scalar from TKW's 1-D vector on the specified index.
+
+    This can also be viewed as indexing/slicing on the fastest
+    dimension. Hence, the semantic of this op is designed to
+    see itself as a reduction on the indexed/fastest dimension.
+    """
+
+    register_: fx.Proxy
+    offset: IndexExpr | int
+
+    @property
+    def type(self) -> "Register":
+        # Intuition here is we are trying to extract an element
+        # from fastest dim => we reduce the fastest dim.
+        src_type = get_custom(self.register_).type
+        # Return itself if just 0-D/1-D symbolic.
+        if len(src_type.symbolic_shape) <= 1:
+            return src_type
+
+        # Typically only fastest dim has non-unit dim,
+        # but if all unit-dim get fastest/last one.
+        all_unit_dims = lambda index: all(x.size == 1 for x in index.values())
+        if all_unit_dims(self.register_.index):
+            dst_shape = src_type.symbolic_shape[:-1]
+            dst_type = Register[*dst_shape, src_type.dtype]
+            return dst_type
+
+        # If non-unit dim exists => non-unit dim is fastest dim.
```

Comment:
okay this is a little non-intuitive from the semantics of the op. Basically, this extract op is a slice along the axis with the most elements rather than the fastest dimension.

Reply:
Alternatively, you can call it a slice in the SIMT/thread world, as opposed to the SIMD/wave world?

Reply:
Yeah exactly! it's a SIMT slice, it is expected to be used as a lower level op. We extract in the vector/post-distributed/SIMT world, but it essentially retains symbolic shape of the whole tensor S.T we can be compliant in higher level analysis.

Reply:
make sense, can you update the documentation to describe that?